rfcs/webtransport_h3_test_server.md

```diff
@@ -0,0 +1,94 @@
+
+This RFC shares the same background/motivation with [RFC 42](https://github.com/web-platform-tests/rfcs/blob/master/rfcs/quic.md) but focuses on WebTransport over HTTP/3. See RFC 42 for more background.
+
+## Details
+
+### Implementation
+
+The WebTransport over HTTP/3 test server is built on top of `aioquic`, which is already added under `tools/third_party`. As of version 0.9.11, `aioquic` lacks some mechanisms to implement a WebTransport server (e.g. configuring SETTINGS HTTP/3 frame) but we plan to update `aioquic` once it supports necessary mechanisms.
+
+The server will be implemented under `tools/webtransport` directory.
+
+### Handlers
+
+A WebTransport handler is a Python script which contains a function `handle_event()`. Here is an example handler which echos back received data.
+
+```python
+from webtransport import (
+    WebTransportSession,
+    BidirectionalStreamDataReceived,
+    DatagramReceived,
+    WebTransportEvent
+)
+
+def handle_event(session: WebTransportSession,
+                 event: WebTransportEvent) -> None:
+    if isinstance(event, DatagramReceived):
+        session.send_datagram(data=event.data)
+    elif isinstance(event, BidirectionalStreamDataReceived):
+        session.send_stream_data(
+            stream_id=event.stream_id, data=event.data)
+```
+
+`session` is an object that represents a WebTransport over HTTP/3 session. It provides APIs to handle the session. Current proposed APIs can be found [here](https://bashi.github.io/wpt-wt-test-server-api/session.html).
+
+`event` represents a WebTransport event such as a reception of stream data. Current proposed events can be found [here](https://bashi.github.io/wpt-wt-test-server-api/events.html).
+
+#### Alternative handler design
+
+Alternatively, we can define a set of functions which are called upon each event. This API pattern is similar to existing [H2 handlers](https://web-platform-tests.org/writing-tests/h2tests.html). The equivalent of the above example might look like:
+
+```python
+def datagram_received(session: WebTransportSession,
+                      data: bytes):
+    session.send_datagram(data=data)
+
+
+def bidirectional_stream_data_received(session: WebTransportSession,
+                                       stream_id: int,
+                                       data: bytes):
+    session.send_stream_data(stream_id_id=stream_id, data=data)
+```
+
+A major downside of this approach is that keeping API documents and implementations in sync would require extra works, as H2 handler API document duplicates comments from the source.
+
+Another downside of this approach is that we won't have editor/IDE supports on what kind of callback functions are available.
+
+TODO: Spend more time on the API design. People may prefer this alternative approach.
```

Comment:
It feels like a seperate RFC is needed for the API design.

Reply:
I decided to use this approach (callback function style) and abandoned the event style approach. I guess we don't need a separate RFC?